### PR TITLE
fungible conformance tests: Inspect and Mutate

### PR DESCRIPTION
**Pull request by liamaharon**
Partial: https://github.com/paritytech/substrate/issues/13252

Implements conformance tests for `fungible` traits `Inspect` and `Mutate`.

- [x] Pattern for creating a function that accepts something implementing `Inspect + Mutate`
    - ```rust
        pub fn something<T, AccountId, Balance>()
        where
	        T: Mutate<AccountId> + Inspect<AccountId, Balance = Balance>,
	        AccountId: AtLeast8BitUnsigned,
	        Balance: AtLeast8BitUnsigned + Debug,
        {}
        ```

- [x] Implement inspect and mutate conformance tests (`frame/support/src/traits/tokens/fungible/conformance_tests/inspect_mutate.rs`)
- [x] Test conformance tests against `balances` (`frame/balances/src/tests/fungible_conformance_tests.rs`)
     - ```
        cargo test --package pallet-balances -- tests::fungible_conformance_tests:: --nocapture
        ```
- [x] Create a macro to generate test cases wrapped with some logic

### CODE REVIEW

File: frame/balances/src/tests/fungible_conformance_tests.rs [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+#[test]
+fn mint_into_success() {
+	ExtBuilder::default().build_and_execute_with(|| {
+		conformance_tests::inspect_mutate::mint_into_success::<
+			Balances,
+			<Test as frame_system::Config>::AccountId,
+			<Test as pallet_balances::Config>::Balance,
+		>();
+	});
+}
```
Comment (ggwpez) [inline frame/balances/src/tests/fungible_conformance_tests.rs:32]:
```suggestion
macro_rules! run_test {
	($name:ident) => {
		#[test]
		fn $name() {
			ExtBuilder::default().build_and_execute_with(|| {
				conformance_tests::inspect_mutate::$name::<
					Balances,
					<Test as frame_system::Config>::AccountId,
					<Test as pallet_balances::Config>::Balance,
				>();
			});
		}	
	}
}

run_test!(mint_into_success);
```

File: frame/support/src/traits/tokens/fungible/conformance_tests/inspect_mutate.rs [hunk 2]
```diff
@@ -0,0 +1,659 @@
...
+	},
+};
+use core::fmt::Debug;
+use sp_arithmetic::traits::AtLeast8BitUnsigned;
+
+//
+// mint_into
+//
+
+pub fn mint_into_success<T, AccountId, Balance>()
```
Comment (ggwpez) [inline frame/support/src/traits/tokens/fungible/conformance_tests/inspect_mutate.rs:31]: I think we could also employ fuzzers - eventually.  
They look quite similar to your pattern so far and test the assumptions with many different inputs.  
A recent one is [this](https://github.com/paritytech/substrate/blob/344d3a6f850ef52e05368fdd58f7d9461658185b/primitives/arithmetic/fuzzer/src/per_thing_mult_fraction.rs#L42) and the `per_thing_from_rational` in that folder.

**Comment by sam0x17** [comment]
This should lend itself well to a simple macro. Based on your description I think you should be able to write a regular proc macro that accepts those args (I generally recommend making a MyUseCaseArgs struct and using `derive_syn_parse` to get an automatic Parse implementation for it, it can even do things like optional fields etc). The proc macro would then expand to a test case like the one you are writing.

you mention that `Balance` and `AccountId` need to be passed. Are these always just `<Test as frame_system::Config>::AccountId` and `<Test as pallet_balances::Config>::Balance` or is there some variation there necessitating them being part of the input. I'm assuming based on your examples that these are always the same?

If the inputs are just a few idents or paths, you could have something like:

```rust
#[derive(derive_syn_parse::Parse)]
struct InspectMutateConformanceTestArgs {
    some_ident: syn::Ident,
    _comma1: syn::token::Comma,
    some_path: syn::Path,
    _comma2: Option<syn::token::Comma>,
    #[parse_if(_comma2.is_some())]
    optional_path: Option<syn::Path>,
}

#[proc_macro]
pub fn inspect_mutate_conformance_tests(tokens: TokenStream) -> TokenStream {
    let args = parse_macro_input!(tokens as InspectMutateConformanceTestArgs);
    let test_ident = quote::format_ident!("mint_into_{}", args.some_ident);
    // other stuff with args ...
    quote::quote! {
        #[test]
        fn #test_ident() {
            ExtBuilder::default().build_and_execute_with(|| {
                conformance_tests::inspect_mutate::#test_ident::<
                    Balances,
                    <Test as frame_system::Config>::AccountId,
                    <Test as pallet_balances::Config>::Balance,
                >();
            });
        }
    }
}
```

Something vaguely like that, I'm just not 100% clear on what the args would be, but the above would let you parse args like this:

```rust
inspect_mutate_conformance_tests!(an_ident, some::cool_path);
```

And optionally have a third field like:

```rust
inspect_mutate_conformance_tests!(another_ident, different::path, an::optional::path);
```

Hope this helps. If you can give me exact desired usage of the proc macro I can give you something tighter

**Comment by KiChjang** [comment]
I think something as simple as this doesn't require a proc macro, and can still be easily done using a declarative macro using `macro_rules!`:

```rust
macro_rules! inspect_mutate_conformance_tests {
    ($name:ident, $some_path:path $(, $optional_path:path)?) => /* macro code goes here */
}
```

And with the clever usage of `paste::paste!`, we can create custom names for the tests as well using declarative macros.

**Comment by sam0x17** [comment]
Yes I agree a macro_rules would be fine here unless you want to be very strict about what you are passing in which looks like it is not necessary if it's idents and paths

**Comment by liamaharon** [comment]
Thanks all, really appreciate the examples.

Kian are I were hoping we wouldn't need to make any assumptions in the macro about an `ExtBuilder` existing, but that is probably too much extra complexity for too little gain at this point.

File: frame/support/src/traits/tokens/fungible/conformance_tests/inspect_mutate.rs (hunk 2 above)
Comment (liamaharon) [reply frame/support/src/traits/tokens/fungible/conformance_tests/inspect_mutate.rs:31]: Fuzzing is a great idea! I'll create an issue to work on that once some initial coverage is done.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658780

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2661075

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2661569

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2661884

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2661883

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2662825

File: frame/balances/src/tests/fungible_conformance_tests.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2023 Parity Technologies (UK) Ltd.
```
Comment (kianenigma) [inline frame/balances/src/tests/fungible_conformance_tests.rs:3]:
```suggestion
// Copyright (C) Parity Technologies (UK) Ltd.
```

File: frame/support/src/traits/tokens/fungible/conformance_tests/inspect_mutate.rs [hunk 4]
```diff
@@ -0,0 +1,734 @@
...
+	},
+};
+use core::fmt::Debug;
+use sp_arithmetic::traits::AtLeast8BitUnsigned;
+
+//
+// mint_into
+//
+
+pub fn mint_into_success<T, AccountId, Balance>(_dust_trap: Option<AccountId>)
```
Comment (kianenigma) [inline frame/support/src/traits/tokens/fungible/conformance_tests/inspect_mutate.rs:31]: FYI, I think soon we should enforce 

```
#![deny(missing_docs)]
#![deny(clippy::missing_docs_in_private_items)]
```

So you might want to think about it already.

File: frame/support/src/traits/tokens/fungible/conformance_tests/inspect_mutate.rs [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+where
+	T: Mutate<AccountId> + Inspect<AccountId, Balance = Balance>,
```
Comment (kianenigma) [inline frame/support/src/traits/tokens/fungible/conformance_tests/inspect_mutate.rs:33]: Could we have not just used `<T as Inspect<AccountId>>::Balance` go the the right Balance type? 

You can still bound it as well: `where <T as Inspect<AccountId>>::Balance: X + Y + Z`

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2663365

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2663590

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2663589

**Review by sam0x17 — APPROVED** [review]

**Review by gavofyork — APPROVED** [review]
Looks good.

File: frame/balances/src/tests/fungible_conformance_tests.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+use frame_support::traits::fungible::{conformance_tests, Inspect, Mutate};
+use paste::paste;
+
+macro_rules! run_tests {
+    ($parent_module:tt :: $child_module:tt, $ext_deposit:expr, $($name:ident),*) => {
```
[the remark this answers is not in the document]
Comment (liamaharon) [reply frame/balances/src/tests/fungible_conformance_tests.rs:23]: thanks 👍

**Pull request merged**